### PR TITLE
fix-locked-samples-on-project-create

### LINKED ISSUE #183: Creating a new project from samples in cart always results in locked samples
Seems similar to #667

Imported from GitLab issue #679.  Originally posted on 2018/09/12 04:23PM

### PR DESCRIPTION
**Pull request by deepsidhu85**
## Description of changes
What did you change in this pull request?  Provide a description of files changed, user interactions changed, etc.  Include how to test your changes.

Added functionality for a project owner to either leave samples unlocked or let them lock the samples from modification from other users (both managers and collaborators) that are members of the project. This is accomplished by adding a "Lock sample modification" checkbox to the Create New Project page

**To test locked samples:**
To test this functionality add samples to the cart and create a new project from these samples. On the create a project page select the "Lock sample modification" checkbox. Create the project. Add another user and add them to this project with a manager project role. Log in as this new user and go to the projects page and view the project that this user was added to. Click on any of the samples in the project which leads you to the sample details page. You should not have the ability to edit this sample.

**To test unlocked samples:**
To test this functionality add samples to the cart and create a new project from these samples. On the create a project page leave the "Lock sample modification" checkbox unchecked. Create the project. Add another user and add them to this project with a manager project role. Log in as this new user and go to the projects page and view the project that this user was added to. Click on any of the samples in the project which leads you to the sample details page. You should have the ability to edit this sample.

**Note:**
To test locked and unlocked samples for a collaborator follow the directions above but when adding the user to the project select **collaborator** for the project role. In either case a collaborator should **not** be allowed to edit any samples

## Related issue
Link to the GitHub issue this pull request addresses using the `#issuenum` format.  If it completes an issue, use `Fixes #issuenum` to automatically close the issue.

Fixes #183

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [x] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
* [x] Tests added (or description of how to test) for any new features.
* [x] User documentation updated for UI or technical changes.

### CODE REVIEW

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectsController.java [hunk 1]
```diff
@@ -233,35 +218,38 @@ public String getProjectSpecificPage(@PathVariable Long projectId, final Model m
...
 	 * @return The name of the create new project page
 	 */
 	@RequestMapping(value = "/projects/new", method = RequestMethod.GET)
 	public String getCreateProjectPage(
 			@RequestParam(name = "cart", required = false, defaultValue = "false") boolean useCartSamples,
-			final Model model) {
+			final Model model,
+			@RequestParam(name = "lockSamples", required = false, defaultValue = "true") boolean owner) {
 		model.addAttribute("useCartSamples", useCartSamples);
+		model.addAttribute("lockSamples", owner);
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectsController.java]: Is this actually used in the view?  I don't see anything using `${lockSamples}` in `project_new.html`

File: src/main/webapp/pages/projects/project_new.html [hunk 2]
```diff
@@ -213,6 +220,25 @@ <h2 class="panel-title" th:text="#{projects.create.heading}">Create a New Projec
           }
         });
 
+        $("#useCartSamples").on("change", function(e) {
```
Comment (tom114) [inline src/main/webapp/pages/projects/project_new.html]: Add a comment above these 2 functions about what they're doing please.

File: src/main/webapp/pages/projects/project_new.html [hunk 3]
```diff
@@ -94,15 +94,22 @@ <h2 class="panel-title" th:text="#{projects.create.heading}">Create a New Projec
...
                                 </div>
-                            </th:block>
-                            
+
                                 <div class="checkbox">
-                                    <label for="assemble">
-                                        <input type="checkbox" id="assemble" name="assembleUploads" value="true"/>
-                                        <span th:text="#{projects.create.settings.assemble}">_Assemble_</span>
+                                    <label for="lockSamples">
+                                        <input type="checkbox" id="lockSamples" name="lockSamples" value="false" th:disabled="${allSamples} == 0" />
```
Comment (tom114) [inline src/main/webapp/pages/projects/project_new.html]: When I go to `/projects/new` both the "Add Samples..." and "Lock sample..." boxes are unchecked.  Shouldn't the lock one be disable at this point?

image

File: src/main/webapp/pages/projects/project_new.html [hunk 4]
```diff
@@ -213,6 +220,25 @@ <h2 class="panel-title" th:text="#{projects.create.heading}">Create a New Projec
...
+            $("#lockSamples").removeAttr("disabled");
+          }
+        });
+
+        $("#lockSamplesToOwner").on("change", function(e) {
+          var selected = $(e.target).prop("checked");
+          if (selected) {
+            $("#lockSamples").val("false");
+          } else {
+            $("#lockSamples").val("true");
```
Comment (tom114) [inline src/main/webapp/pages/projects/project_new.html]: Why flip the value of the checkbox?  If it's unchecked it just wouldn't send the value to the server on the project create post.

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectsController.java [hunk 5]
```diff
@@ -125,8 +125,7 @@
 
 	/*
 	 * Converters
-	 */
-	Formatter<Date> dateFormatter;
+	 */ Formatter<Date> dateFormatter;
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectsController.java]: Hey. I have the file loaded into intellij. For some reason when I run the formatter on that block it goes right back to `*/ Formatter<Date> dateFormatter;`

[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectsController.java]: Fixed in 9e258b0be30f71fb335b7baf9d4591ba3e44601b

File: src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectsController.java (hunk 1 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/ria/web/projects/ProjectsController.java]: Fixed in 9e258b0be30f71fb335b7baf9d4591ba3e44601b

File: src/main/webapp/pages/projects/project_new.html (hunk 2 above)
Comment (deepsidhu85) [reply src/main/webapp/pages/projects/project_new.html]: Fixed in 9e258b0be30f71fb335b7baf9d4591ba3e44601b

File: src/main/webapp/pages/projects/project_new.html (hunk 3 above)
Comment (deepsidhu85) [reply src/main/webapp/pages/projects/project_new.html]: Fixed in 9e258b0be30f71fb335b7baf9d4591ba3e44601b

File: src/main/webapp/pages/projects/project_new.html (hunk 4 above)
Comment (deepsidhu85) [reply src/main/webapp/pages/projects/project_new.html]: Fixed in 9e258b0be30f71fb335b7baf9d4591ba3e44601b

File: src/main/webapp/pages/projects/project_new.html [hunk 6]
```diff
@@ -213,6 +220,32 @@ <h2 class="panel-title" th:text="#{projects.create.heading}">Create a New Projec
...
+        /*
+            Disables/enables the lockSamples checkbox
+            depending on if using the samples from cart
+        */
+        $("#useCartSamples").on("change", function(e) {
+          var selected = $(e.target).prop("checked");
+          if (!selected) {
+            $("#lockSamples").attr("disabled", true);
+          } else {
+            $("#lockSamples").removeAttr("disabled");
```
Comment (tom114) [inline src/main/webapp/pages/projects/project_new.html]: Add a `$("#lockSamples").prop("checked", false)` above when you disable the checkbox to automatically uncheck it too please.

File: src/main/webapp/pages/projects/project_new.html [hunk 7]
```diff
@@ -213,6 +220,32 @@ <h2 class="panel-title" th:text="#{projects.create.heading}">Create a New Projec
...
+
+        /*
+            If lockSamples checkbox is checked then we set
+            it's value to false which will set
+            the sample owner to false
+        */
+        $("#lockSamples").on("change", function(e) {
+          var selected = $(e.target).prop("checked");
+          if (selected) {
+            $("#lockSamples").val("false");
```
Comment (tom114) [inline src/main/webapp/pages/projects/project_new.html]: You can test to be sure, but with the `@RequestParam(name = "lockSamples", required = false, defaultValue = "true") boolean owner` on the controller i'm pretty sure you don't need this at all.  If the checkbox is checked, `owner` will always be set to false, if it's unchecked it'll by default be true.

File: src/main/webapp/pages/projects/project_new.html (hunk 6 above)
Comment (deepsidhu85) [reply src/main/webapp/pages/projects/project_new.html]: Fixed in 94d58a64cfc89091be01471a00966a4ad200ad8e

File: src/main/webapp/pages/projects/project_new.html (hunk 7 above)
Comment (deepsidhu85) [reply src/main/webapp/pages/projects/project_new.html]: Fixed in 94d58a64cfc89091be01471a00966a4ad200ad8e

**Pull request merged**